File: recDP/002-fib3.py
def fibo(n):
    curr = 0
    next = 1 # wasBUG: change 2 to 1 coz curr starts from 0
    for _ in range(1, n + 1): # Bug start from 2
        next_next = curr + next
        curr = next
        next = next_next
    return curr

print(fibo(6)) # 8
print(fibo(4)) # 3

# overlapping (sub)problems
# optimal (sub)structure

#           fib(8)
#   fib(7)         fib(6)
# fib(6) fib(5)    .  .

# first write the general case i.e. the recurrence breakdown: assume you'll get the lower answers and asssume you have given the base cases
#   then, write all the base cases
#     then, execute(execution context,etc) and check
#       [then, the O(T(n)), O(S(n))]

File: old/recDP/002-fib3.cpp
#include <bits/stdc++.h>
using namespace std;


int fib(int n) {
    int curr = 0;
    int next = 1;
    for (int i = 1; i <= n; i++) {
        int next_next = curr + next;
        curr = next;
        next = next_next; 
    }

    return curr;
}

int main(void) {
    cout << fib(5) << endl;
    cout << fib(6) << endl;
    cout << fib(8) << endl;

    int x;
    cin >> x;
    cout << x << endl;
}
